File: generated-code/csharp/codecraft/FileReadWrite/Model/DebugInterface/DebugData.cs
namespace TransGenTest.Model.DebugInterface
{
    /// <summary>
    /// Debug data can be drawn in the app
    /// </summary>
    public abstract class DebugData
    {
        /// <summary> Write DebugData to writer </summary>
        public abstract void WriteTo(System.IO.BinaryWriter writer);

        /// <summary> Read DebugData from reader </summary>
        public static DebugData ReadFrom(System.IO.BinaryReader reader)
        {
            switch (reader.ReadInt32())
            {
                case Log.TAG:
                    return Log.ReadFrom(reader);
                case Primitives.TAG:
                    return Primitives.ReadFrom(reader);
                case PlacedText.TAG:
                    return PlacedText.ReadFrom(reader);
                default:
                    throw new System.Exception("Unexpected tag value");
            }
        }

        /// <summary>
        /// Log some text
        /// </summary>
        public class Log : DebugData
        {
            public const int TAG = 0;
        
            /// <summary>
            /// Text to show
            /// </summary>
            public string Text { get; set; }
        
            public Log() { }
        
            public Log(string text)
            {
                this.Text = text;
            }
        
            /// <summary> Read Log from reader </summary>
            public static new Log ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new Log();
                result.Text = System.Text.Encoding.UTF8.GetString(reader.ReadBytes(reader.ReadInt32()));
                return result;
            }
        
            /// <summary> Write Log to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
                var textData = System.Text.Encoding.UTF8.GetBytes(Text);
                writer.Write(textData.Length);
                writer.Write(textData);
            }
        
            /// <summary> Get string representation of Log </summary>
            public override string ToString() {
                string stringResult = "Log { ";
                stringResult += "Text: ";
                stringResult += "\"" + Text + "\"";
                stringResult += " }";
                return stringResult;
            }
        }

        /// <summary>
        /// Draw primitives
        /// </summary>
        public class Primitives : DebugData
        {
            public const int TAG = 1;
        
            /// <summary>
            /// Vertices
            /// </summary>
            public Model.DebugInterface.ColoredVertex[] Vertices { get; set; }
            /// <summary>
            /// Primitive type
            /// </summary>
            public Model.DebugInterface.PrimitiveType PrimitiveType { get; set; }
        
            public Primitives() { }
        
            public Primitives(Model.DebugInterface.ColoredVertex[] vertices, Model.DebugInterface.PrimitiveType primitiveType)
            {
                this.Vertices = vertices;
                this.PrimitiveType = primitiveType;
            }
        
            /// <summary> Read Primitives from reader </summary>
            public static new Primitives ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new Primitives();
                result.Vertices = new Model.DebugInterface.ColoredVertex[reader.ReadInt32()];
                for (int verticesIndex = 0; verticesIndex < result.Vertices.Length; verticesIndex++)
                {
                    result.Vertices[verticesIndex] = Model.DebugInterface.ColoredVertex.ReadFrom(reader);
                }
                result.PrimitiveType = PrimitiveTypeHelper.ReadFrom(reader);
                return result;
            }
        
            /// <summary> Write Primitives to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
                writer.Write(Vertices.Length);
                foreach (var verticesElement in Vertices)
                {
                    verticesElement.WriteTo(writer);
                }
                writer.Write((int) (PrimitiveType));
            }
        
            /// <summary> Get string representation of Primitives </summary>
            public override string ToString() {
                string stringResult = "Primitives { ";
                stringResult += "Vertices: ";
                stringResult += "[ ";
                int verticesIndex = 0;
                foreach (var verticesElement in Vertices)
                {
                    if (verticesIndex != 0) {
                        stringResult += ", ";
                    }
                    stringResult += verticesElement.ToString();
                    verticesIndex++;
                }
                stringResult += " ]";
                stringResult += ", ";
                stringResult += "PrimitiveType: ";
                stringResult += PrimitiveType.ToString();
                stringResult += " }";
                return stringResult;
            }
        }

        /// <summary>
        /// Draw text
        /// </summary>
        public class PlacedText : DebugData
        {
            public const int TAG = 2;
        
            /// <summary>
            /// Vertex to determine text position and color
            /// </summary>
            public Model.DebugInterface.ColoredVertex Vertex { get; set; }
            /// <summary>
            /// Text
            /// </summary>
            public string Text { get; set; }
            /// <summary>
            /// Text alignment (0 means left, 0.5 means center, 1 means right)
            /// </summary>
            public float Alignment { get; set; }
            /// <summary>
            /// Font size in pixels
            /// </summary>
            public float Size { get; set; }
        
            public PlacedText() { }
        
            public PlacedText(Model.DebugInterface.ColoredVertex vertex, string text, float alignment, float size)
            {
                this.Vertex = vertex;
                this.Text = text;
                this.Alignment = alignment;
                this.Size = size;
            }
        
            /// <summary> Read PlacedText from reader </summary>
            public static new PlacedText ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new PlacedText();
                result.Vertex = Model.DebugInterface.ColoredVertex.ReadFrom(reader);
                result.Text = System.Text.Encoding.UTF8.GetString(reader.ReadBytes(reader.ReadInt32()));
                result.Alignment = reader.ReadSingle();
                result.Size = reader.ReadSingle();
                return result;
            }
        
            /// <summary> Write PlacedText to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
                Vertex.WriteTo(writer);
                var textData = System.Text.Encoding.UTF8.GetBytes(Text);
                writer.Write(textData.Length);
                writer.Write(textData);
                writer.Write(Alignment);
                writer.Write(Size);
            }
        
            /// <summary> Get string representation of PlacedText </summary>
            public override string ToString() {
                string stringResult = "PlacedText { ";
                stringResult += "Vertex: ";
                stringResult += Vertex.ToString();
                stringResult += ", ";
                stringResult += "Text: ";
                stringResult += "\"" + Text + "\"";
                stringResult += ", ";
                stringResult += "Alignment: ";
                stringResult += Alignment.ToString();
                stringResult += ", ";
                stringResult += "Size: ";
                stringResult += Size.ToString();
                stringResult += " }";
                return stringResult;
            }
        }
    }
}

File: generated-code/php/codecraft/FileReadWrite/Model/DebugInterface/DebugData.php
<?php

namespace Model\DebugInterface {

    require_once 'Model/DebugInterface/ColoredVertex.php';
    require_once 'Model/DebugInterface/PrimitiveType.php';

    /**
     * Debug data can be drawn in the app
     */
    abstract class DebugData
    {
        /**
         * Write DebugData to output stream
         */
        abstract function writeTo($stream);

        /**
         * Read DebugData from input stream
         */
        static function readFrom($stream)
        {
            $tag = $stream->readInt32();
            if ($tag == \Model\DebugInterface\DebugData\Log::TAG) {
                return \Model\DebugInterface\DebugData\Log::readFrom($stream);
            }
            if ($tag == \Model\DebugInterface\DebugData\Primitives::TAG) {
                return \Model\DebugInterface\DebugData\Primitives::readFrom($stream);
            }
            if ($tag == \Model\DebugInterface\DebugData\PlacedText::TAG) {
                return \Model\DebugInterface\DebugData\PlacedText::readFrom($stream);
            }
            throw new Exception('Unexpected tag value');
        }
    }
}

namespace Model\DebugInterface\DebugData {
    /**
     * Log some text
     */
    class Log extends \Model\DebugInterface\DebugData
    {
        const TAG = 0;
    
        /**
         * Text to show
         */
        public $text;
    
        function __construct($text)
        {
            $this->text = $text;
        }
    
        /**
         * Read Log from input stream
         */
        public static function readFrom($stream)
        {
            $text = $stream->readString();
            return new Log($text);
        }
    
        /**
         * Write Log to output stream
         */
        public function writeTo($stream)
        {
            $stream->writeInt32(Log::TAG);
            $stream->writeString($this->text);
        }
    }

    /**
     * Draw primitives
     */
    class Primitives extends \Model\DebugInterface\DebugData
    {
        const TAG = 1;
    
        /**
         * Vertices
         */
        public $vertices;
        /**
         * Primitive type
         */
        public $primitiveType;
    
        function __construct($vertices, $primitiveType)
        {
            $this->vertices = $vertices;
            $this->primitiveType = $primitiveType;
        }
    
        /**
         * Read Primitives from input stream
         */
        public static function readFrom($stream)
        {
            $vertices = [];
            $verticesSize = $stream->readInt32();
            for ($verticesIndex = 0; $verticesIndex < $verticesSize; $verticesIndex++) {
                $verticesElement = \Model\DebugInterface\ColoredVertex::readFrom($stream);
                $vertices[] = $verticesElement;
            }
            $primitiveType = \Model\DebugInterface\PrimitiveType::readFrom($stream);
            return new Primitives($vertices, $primitiveType);
        }
    
        /**
         * Write Primitives to output stream
         */
        public function writeTo($stream)
        {
            $stream->writeInt32(Primitives::TAG);
            $stream->writeInt32(count($this->vertices));
            foreach ($this->vertices as $element) {
                $element->writeTo($stream);
            }
            $stream->writeInt32($this->primitiveType);
        }
    }

    /**
     * Draw text
     */
    class PlacedText extends \Model\DebugInterface\DebugData
    {
        const TAG = 2;
    
        /**
         * Vertex to determine text position and color
         */
        public $vertex;
        /**
         * Text
         */
        public $text;
        /**
         * Text alignment (0 means left, 0.5 means center, 1 means right)
         */
        public $alignment;
        /**
         * Font size in pixels
         */
        public $size;
    
        function __construct($vertex, $text, $alignment, $size)
        {
            $this->vertex = $vertex;
            $this->text = $text;
            $this->alignment = $alignment;
            $this->size = $size;
        }
    
        /**
         * Read PlacedText from input stream
         */
        public static function readFrom($stream)
        {
            $vertex = \Model\DebugInterface\ColoredVertex::readFrom($stream);
            $text = $stream->readString();
            $alignment = $stream->readFloat32();
            $size = $stream->readFloat32();
            return new PlacedText($vertex, $text, $alignment, $size);
        }
    
        /**
         * Write PlacedText to output stream
         */
        public function writeTo($stream)
        {
            $stream->writeInt32(PlacedText::TAG);
            $this->vertex->writeTo($stream);
            $stream->writeString($this->text);
            $stream->writeFloat32($this->alignment);
            $stream->writeFloat32($this->size);
        }
    }
}